typescript/lib/callback.ts

```diff
@@ -0,0 +1,97 @@
+import { createLogger, format, transports } from "winston";
+
+// Constants
+const DEFAULT_TIMEOUT: number = 5; // Default timeout for callback execution in seconds
+
+export class CallbackEvent {
+  name: string;
+  file: string;
+  data: any;
```

Comment:
Not a big fan of using `any` in ts in general (also, surprised the linter didn't complain), and in this case I think it does result in extra friction on people using callbacks since they need to dig through code to see what possible data values will be for different event types.

Not blocking on this, but it seems that there's a pretty limited set of events that are defined and could just do something like

```
export interface CallbackEvent<D=unknown> {
  name: string,
  data: D,
  ts_ns?: number,
}

// In ModelParser

export interface ModelParserEvent<D=unknown> extends CallbackEvent<D> {
  parser: string; // "openai", etc.
};

export type SerializeStartEventData<SD=JSONObject> = {
  promptName: string,
  data: SD,
  params: JSONObject,
};

export interface SerializeStartEvent<SD=JSONObject> extends ModelParserEvent<
  SerializeStartEventData<SD>
> {
  name: "on_serialize_start",
};

export interface SerializeEndEvent extends ModelParserEvent<Prompt[]> {
  name: "on_serialize_end"
};

export interface DeserializeStartEvent extends ModelParserEvent<
  {prompt: Prompt, params?: JSONObject}
> {name: "on_deserialize_start"};

export interface DeserializeEndEvent<R = JSONObject> extends ModelParserEvent<{result: R}> {
   name: "on_deserialize_end"
};

export interface RunStartEvent extends ModelParserEvent<{
  prompt: Prompt,
  options?: InferenceOptions,
  params?: JSONObject,
}> {
  name: "on_run_start",
};

export interface RunEndEvent extends ModelParserEvent<{
  result: Output | Output[],
}> {
  name: "on_run_end",
};
```

Then we could leverage stronger typing of data while still supporting arbitrary events,

Reply:
Also just want to reiterate that this isn't a blocker here, just something that can be looked into later to improve the typing a bit

Reply:
adding to linear todo to revisit